### PR TITLE
Calibrate models

### PR DESCRIPTION
**Pull request by meelislootus**

### CODE REVIEW

**Comment by Mi1ind** [comment]
Current compile issue:
It can be replicated by running `cargo test` in the `proc_block/fft` subdirectory


```
error[E0308]: mismatched types
  --> proc_blocks/fft/src/lib.rs:78:59
   |
78 |         let mut mel_filter_matrix = DMatrix::<f64>::zeros(filter_count, power_spectrum_size);
   |                                                           ^^^^^^^^^^^^ expected `usize`, found `i32`
   |
help: you can convert an `i32` to a `usize` and panic if the converted value doesn't fit
   |
78 |         let mut mel_filter_matrix = DMatrix::<f64>::zeros(filter_count.try_into().unwrap(), power_spectrum_size);
   |                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> proc_blocks/fft/src/lib.rs:80:13
   |
80 |             self.sample_rate,
   |             ^^^^^^^^^^^^^^^^ expected `usize`, found `u32`
   |
help: you can convert a `u32` to a `usize` and panic if the converted value doesn't fit
   |
80 |             self.sample_rate.try_into().unwrap(),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
  --> proc_blocks/fft/src/lib.rs:83:13
   |
83 |             filter_count,
   |             ^^^^^^^^^^^^ expected `usize`, found `i32`
   |
help: you can convert an `i32` to a `usize` and panic if the converted value doesn't fit
   |
83 |             filter_count.try_into().unwrap(),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `f32: From<&f64>` is not satisfied
   --> proc_blocks/fft/src/lib.rs:96:32
    |
96  |         let mel_spectrum_f32 = mel_spectrum_matrix.into_iter().map(f32::from);
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<&f64>` is not implemented for `f32`
    | 
   ::: /Users/milind/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:365:1
    |
365 | pub trait From<T>: Sized {
    | ------------------------ required by this bound in `From`
    |
    = help: the following implementations were found:
              <f32 as From<i16>>
              <f32 as From<i8>>
              <f32 as From<u16>>
              <f32 as From<u8>>

error[E0277]: the trait bound `f32: From<&f64>` is not satisfied
  --> proc_blocks/fft/src/lib.rs:96:68
   |
96 |         let mel_spectrum_f32 = mel_spectrum_matrix.into_iter().map(f32::from);
   |                                                                    ^^^^^^^^^ the trait `From<&f64>` is not implemented for `f32`
   |
   = help: the following implementations were found:
             <f32 as From<i16>>
             <f32 as From<i8>>
             <f32 as From<u16>>
             <f32 as From<u8>>
   = note: required by `from`

error[E0599]: no method named `iter` found for struct `Map<MatrixIter<'_, f64, Dynamic, Dynamic, VecStorage<f64, Dynamic, Dynamic>>, fn(&f64) -> f32 {<f32 as From<&f64>>::from}>` in the current scope
  --> proc_blocks/fft/src/lib.rs:98:42
   |
98 |         let min_value = mel_spectrum_f32.iter().fold(f32::INFINITY, |a, &b| a.min(b));
   |                                          ^^^^ private field, not a method

error[E0599]: no method named `iter` found for struct `Map<MatrixIter<'_, f64, Dynamic, Dynamic, VecStorage<f64, Dynamic, Dynamic>>, fn(&f64) -> f32 {<f32 as From<&f64>>::from}>` in the current scope
   --> proc_blocks/fft/src/lib.rs:100:26
    |
100 |         mel_spectrum_f32.iter().fold(f32::NEG_INFINITY, |a, &b| a.max(b));
    |                          ^^^^ private field, not a method

error[E0277]: the trait bound `f32: From<&f64>` is not satisfied
   --> proc_blocks/fft/src/lib.rs:102:28
    |
102 |           let res: Vec<u8> = mel_spectrum_f32
    |  ____________________________^
103 | |             .into_iter()
    | |________________________^ the trait `From<&f64>` is not implemented for `f32`
    | 
   ::: /Users/milind/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:365:1
    |
365 |   pub trait From<T>: Sized {
    |   ------------------------ required by this bound in `From`
    |
    = help: the following implementations were found:
              <f32 as From<i16>>
              <f32 as From<i8>>
              <f32 as From<u16>>
              <f32 as From<u8>>

error[E0277]: the trait bound `f32: From<&f64>` is not satisfied
   --> proc_blocks/fft/src/lib.rs:102:28
    |
102 |           let res: Vec<u8> = mel_spectrum_f32
    |  ____________________________^
103 | |             .into_iter()
104 | |             .map(|freq| 255.0 * (freq - min_value) / (max_value - min_value))
    | |_____________________________________________________________________________^ the trait `From<&f64>` is not implemented for `f32`
    | 
   ::: /Users/milind/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:365:1
    |
365 |   pub trait From<T>: Sized {
    |   ------------------------ required by this bound in `From`
    |
    = help: the following implementations were found:
              <f32 as From<i16>>
              <f32 as From<i8>>
              <f32 as From<u16>>
              <f32 as From<u8>>

error[E0277]: the trait bound `f32: From<&f64>` is not satisfied
   --> proc_blocks/fft/src/lib.rs:102:28
    |
102 |           let res: Vec<u8> = mel_spectrum_f32
    |  ____________________________^
103 | |             .into_iter()
104 | |             .map(|freq| 255.0 * (freq - min_value) / (max_value - min_value))
105 | |             .map(|freq| freq as u8)
    | |___________________________________^ the trait `From<&f64>` is not implemented for `f32`
    | 
   ::: /Users/milind/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:365:1
    |
365 |   pub trait From<T>: Sized {
    |   ------------------------ required by this bound in `From`
    |
    = help: the following implementations were found:
              <f32 as From<i16>>
              <f32 as From<i8>>
              <f32 as From<u16>>
              <f32 as From<u8>>

error[E0277]: the trait bound `f32: From<&f64>` is not satisfied
   --> proc_blocks/fft/src/lib.rs:102:28
    |
102 |           let res: Vec<u8> = mel_spectrum_f32
    |  ____________________________^
103 | |             .into_iter()
104 | |             .map(|freq| 255.0 * (freq - min_value) / (max_value - min_value))
105 | |             .map(|freq| freq as u8)
106 | |             .collect();
    | |______________________^ the trait `From<&f64>` is not implemented for `f32`
    | 
   ::: /Users/milind/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:365:1
    |
365 |   pub trait From<T>: Sized {
    |   ------------------------ required by this bound in `From`
    |
    = help: the following implementations were found:
              <f32 as From<i16>>
              <f32 as From<i8>>
              <f32 as From<u16>>
              <f32 as From<u8>>

error: aborting due to 11 previous errors; 6 warnings emitted

Some errors have detailed explanations: E0277, E0308, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `fft`
```

File: proc_blocks/fft/src/lib.rs [hunk 1]
```diff
@@ -49,24 +49,62 @@ impl Fft {
...
+        let min_value = mel_spectrum_matrix
+            .data
+            .as_vec()
+            .iter()
+            .fold(f64::INFINITY, |a, &b| a.min(b));
+        let max_value = mel_spectrum_matrix
+            .data
+            .as_vec()
+            .iter()
+            .fold(f64::NEG_INFINITY, |a, &b| a.max(b));
```
Comment (Michael-F-Bryan) [inline proc_blocks/fft/src/lib.rs:101]: It looks like `nalgebra::base::Matrix` has `min()` and `max()` methods.

```suggestion
        let min_value = mel_spectrum_matrix.min();
        let max_value = mel_spectrum_matrix.max();
```

File: proc_blocks/fft/src/lib.rs [hunk 2]
```diff
@@ -49,24 +49,62 @@ impl Fft {
...
+
+        let res: Vec<u8> = mel_spectrum_matrix
+            .data
+            .as_vec()
+            .iter()
+            .map(|freq| 25.0 * (freq - min_value) / (max_value - min_value))
```
Comment (Michael-F-Bryan) [inline proc_blocks/fft/src/lib.rs:107]: I'm not sure how much it helps, but `nalgebra::base::Matrix` has [a `map()` method](https://docs.rs/nalgebra/0.26.1/nalgebra/base/struct.Matrix.html#method.map) for converting one matrix into another by applying a function to each element.

We could probably use it to rewrite the rest of the function as:

```rust
let scaled_spectrum = mel_spectrum_matrix
    .map(|freq| (25.0 * (freq - min_value) / (max_value - min_value)) as u8);

let mut out = [0; 1960];
let spectrum_data = mel_spectrum_matrix.data.as_vec();
out.copy_from_slice(&spectrum_data[..1960]);

out
```

File: proc_blocks/fft/src/lib.rs [hunk 3]
```diff
@@ -49,24 +49,62 @@ impl Fft {
...
+        let power_spectrum_matrix_unflipped: DMatrix<f64> =
+            DMatrix::from_iterator(49, power_spectrum_size, doubles);
+        let power_spectrum_matrix_transposed =
+            power_spectrum_matrix_unflipped.transpose();
+        let mut power_spectrum_vec: Vec<_> =
+            power_spectrum_matrix_transposed.row_iter().collect();
+        &power_spectrum_vec.reverse();
+        let power_spectrum_matrix: DMatrix<f64> =
+            DMatrix::from_rows(&power_spectrum_vec);
+        let mel_spectrum_matrix = &mel_filter_matrix * &power_spectrum_matrix;
```
Comment (Michael-F-Bryan) [inline proc_blocks/fft/src/lib.rs:90]: I'm not going to lose sleep over it, but it'd be nice if there was a more concise way of saying *"multiply `mel_filter_matrix` by the flipped, transposed `spectrogram_f32`"*.

File: proc_blocks/fft/src/lib.rs [hunk 4]
```diff
@@ -1,22 +1,22 @@
 #![no_std]
 
+extern crate std;
```
Comment (Michael-F-Bryan) [inline proc_blocks/fft/src/lib.rs:3]: Do we need this for anything?

We'll be compiling it to `wasm32-unknown-unknown` which doesn't have a standard library (bare WebAssembly doesn't even know it is running on an OS!) so you can't use `std` in this crate.

If it's just for running tests on the host machine you'll conditionally compile it in with something like this:

```rust
#[cfg(test)]
#[macro_use]
extern crate std;
```

**Review by Michael-F-Bryan — COMMENTED** [review]
Nice work wiring up the `mel` crate  so we can implement a STFT!

Are you able to write some tests which make sure we accurately capture various properties of the STFT? For example, when I was implementing a plain FFT I used samples of all zeroes, a step impulse, and a sine wave of known frequency, and made sure the largest value was in a specific bin.

Is the code inside `examples/microspeech/tf_speech_commands/` still needed?

**Comment by Michael-F-Bryan** [comment]
We were having issues with one of `mel`'s transitive dependencies. 

image(1)

Quoting myself from Slack:

> I figured out what the issue is... the `mel` crate pulls in `apodize` which pulls in `nalgebra v0.5.1` (currently at `v0.26.1`) which pulls in `num v0.1.42` (currently at `v0.4.0`), which finally pulls in `rustc_serialize v0.3.24`... which make me think `rustc-serialize` uses conditional compilation to implement a trait differently depending on the target OS/architecture, and because it's been deprecated since 2017 it never got any WebAssembly implementation.

DeepinScreenshot_select-area_20210424010352

This was resolved by forking `mel` to the `hotg-ai` organisation and updating dependencies in `Cargo.toml` - https://github.com/hotg-ai/mel/commit/3c366e47e95e8c6b586fae720bb4a29516dc7fc5

**Pull request merged**